### PR TITLE
[Perf] Optimize video postprocess with pinned DtoH fast path

### PR DESCRIPTION
**Pull request by mglyn**
[Perf] Optimize video finalization before D2H

## Summary

Large decoded videos spend a measurable tail on denormalization, layoutconversion, FP32 materialization, and D2H transfer.

Async pinned D2H and overlap with the next request are already provided by #4981, merged by @BLANKETusers on July 29, 2026 UTC. This PR does not add second stream, staging allocator, or output-transport protocol.

It keeps only the complementary optimization: a small consuming helper performs denormalization, clamping, BCTHW → BTHWC conversion, and FP32 materialization on the producing device before handing the result to the existing Worker IPC path.

The experimental integration is intentionally limited to:

- the shared LTX runtime;
- Wan T2V when frame interpolation is disabled.

Other output types and Wan frame interpolation keep their existing paths. The tradeoff is lower latency at the cost of a temporary output-sized GPU allocation.

## Test Result

Offline `Omni.generate`, eager single-GPU execution, CPU offload disabled, same loaded model/request/seed, one warmup plus five measured runs. Latency includes the final NumPy-ready output boundary.

| Model | Resolution / frames | Before | After | Saving |
| --- | ---: | ---: | ---: | ---: |
| LTX | 1024×576×81 | 4765.88 ms | 4504.16 ms | 261.71 ms (5.49%) |
| LTX | 1536×1024×121 | 21756.65 ms | 20659.37 ms | 1097.29 ms (5.04%) |
| Wan | 1280×720×81 | 16758.47 ms | 16268.51 ms | 489.97 ms (2.92%) |

All sampled outputs were bitwise equal. Pre-commit passed, and the existing LTX and Wan tests passed (`89 passed`).

The remaining review question is the latency versus temporary GPU-memory tradeoff, rather than introducing another asynchronous D2H implementation.

### CODE REVIEW

File: tests/diffusion/test_video_postprocess_stager.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import torch
+
+from vllm_omni.diffusion.postprocess.video import VideoPostprocessStager, _PinnedHostArena
+
+
```
Comment (yenuo26) [inline tests/diffusion/test_video_postprocess_stager.py]: please add mark

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
This branch no longer merges cleanly with main. Please rebase and resolve the conflicts before review.

File: vllm_omni/diffusion/postprocess/video.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+        self._finalizer.detach()
+        if self._output_type == "np":
+            weakref.finalize(result, self._stager._release_lease, self._lease)
+            self._result = result
+            return result
+
+        pil_result = [self._video_processor.numpy_to_pil(batch) for batch in result]
+        self._stager._release_lease(self._lease)
+        self._result = pil_result
+        return pil_result
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/postprocess/video.py]: After self._finalizer.detach(), the transfer's safety-net finalizer is gone. If numpy_to_pil raises (for "pil" output) or weakref.finalize fails (for "np" output, very unlikely), no finalizer remains to release the lease. The pinned memory leaks until pipeline shutdown (stager is tied to ideo_processor lifetime).

File: vllm_omni/diffusion/postprocess/video.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+
+
+def _prepare_video_for_fast_postprocess(
+    video: torch.Tensor,
+    video_processor: VideoProcessor,
+) -> torch.Tensor:
+    if getattr(getattr(video_processor, "config", None), "do_normalize", True):
+        video.mul_(0.5).add_(0.5).clamp_(0, 1)
+    # Input is [B, C, T, H, W]; diffusers returns [B, T, H, W, C].
+    return video.permute(0, 2, 3, 4, 1).to(dtype=torch.float32, copy=True)
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/postprocess/video.py:22]: ul_/add_/clamp_ modify the input tensor (VAE decode output) in-place before the copy is made. Currently safe because VAE decode creates a fresh tensor each call, but this is a fragile contract — if the pipeline is ever refactored to cache or share the decode output, this would silently corrupt data. Consider documenting this contract or using non-in-place ops.

**Comment by mglyn** [comment]
The asynchronous pinned D2H path has already been implemented and merged by @BLANKETusers (zzh) in #4981 on July 29, 2026 UTC (July 30 in UTC+8). It provides the background output thread, dedicated CUDA D2H stream, pinned host buffers, and overlap between output transfer and the next request.

I have therefore removed the duplicate async D2H implementation from this PR.

This PR now keeps only the potentially useful complementary part: consuming the decoded video tensor to perform denormalization, clamping, BCTHW → BTHWC layout conversion, and FP32 materialization on the producing device before handing it to the existing Worker IPC path.

To keep the proposal experimental and reviewable, the final diff only includes minimal integrations for:

- LTX, through its shared runtime
- Wan T2V, while preserving the existing frame-interpolation fallback

MiniMax H3 and Cosmos were tested in the broader local prototype, but their integrations are not included because they require additional model-specific output semantics.

### Benchmark methodology

- Offline `Omni.generate`
- Fixed Worker IPC path
- Eager mode, CPU offload disabled
- Same loaded model, request, and seed
- 1 warmup + 5 measured runs
- End-to-end latency measured through the final NumPy-ready output boundary

| Model | Resolution / frames | Before | After | Saving |
|---|---:|---:|---:|---:|
| LTX | 1024×576×81 | 4765.88 ms | 4504.16 ms | 261.71 ms (5.49%) |
| LTX | 1536×1024×121 | 21756.65 ms | 20659.37 ms | 1097.29 ms (5.04%) |
| Wan | 1280×720×81 | 16758.47 ms | 16268.51 ms | 489.97 ms (2.92%) |
| MiniMax H3, prototype only | 1344×768×209 | 29083.74 ms | 28820.14 ms | 263.60 ms (0.91%) |
| Cosmos, prototype only | 1280×720×189 | 15465.94 ms | 14616.92 ms | 849.02 ms (5.49%) |

All sampled outputs were bitwise equal between the two paths.

The remaining question is therefore the latency versus temporary GPU-memory tradeoff, rather than whether another asynchronous D2H implementation should be added here.

**Comment by mglyn** [comment]
Newest related work: #6212 #6541 #6615.
